app/javascript/mastodon/reducers/contexts.ts

```diff
@@ -0,0 +1,156 @@
+import type {
+  ApiStatusJSON,
+  ApiContextJSON,
+} from 'mastodon/api_types/statuses';
+import type { Status } from 'mastodon/models/status';
+
+import { blockAccountSuccess, muteAccountSuccess } from '../actions/accounts';
+import { fetchContext } from '../actions/statuses';
+import { TIMELINE_UPDATE } from '../actions/timelines';
+import { compareId } from '../compare_id';
+
+interface State {
+  inReplyTos: Record<string, string | undefined>;
+  replies: Record<string, string[] | undefined>;
+}
+
+const initialState: State = {
+  inReplyTos: {},
+  replies: {},
+};
+
+const normalizeContext = (
+  state: State,
+  id: string,
+  { ancestors, descendants }: ApiContextJSON,
+) => {
+  const replies = { ...state.replies };
+  const inReplyTos = { ...state.inReplyTos };
+
+  const addReply = ({
+    id,
+    in_reply_to_id,
+  }: {
+    id: string;
+    in_reply_to_id?: string;
+  }) => {
+    if (!in_reply_to_id) {
+      return;
+    }
+
+    if (!inReplyTos[id]) {
+      const siblings = (replies[in_reply_to_id] ??= []);
+      const index = siblings.findIndex((sibling) => compareId(sibling, id) < 0);
+      siblings.splice(index + 1, 0, id);
+      inReplyTos[id] = in_reply_to_id;
+    }
+  };
+
+  // We know in_reply_to_id of statuses but `id` itself.
+  // So we assume that the status of the id replies to last ancestors.
+
+  ancestors.forEach(addReply);
+
+  if (ancestors[0]) {
+    addReply({
+      id,
+      in_reply_to_id: ancestors[ancestors.length - 1]?.id,
+    });
+  }
+
+  descendants.forEach(addReply);
+
+  return { ...state, replies, inReplyTos };
+};
+
+const deleteFromContexts = (state: State, ids: string[]) => {
+  const replies = { ...state.replies };
+  const inReplyTos = { ...state.inReplyTos };
+
+  ids.forEach((id) => {
+    const inReplyToIdOfId = inReplyTos[id];
+    const repliesOfId = replies[id];
+
+    if (inReplyToIdOfId) {
+      const siblings = replies[inReplyToIdOfId];
+
+      if (siblings) {
+        replies[inReplyToIdOfId] = siblings.filter((sibling) => sibling !== id);
+      }
+    }
+
+    if (repliesOfId) {
+      repliesOfId.forEach((reply) => {
+        inReplyTos[reply] = undefined;
+      });
+    }
+
+    inReplyTos[id] = undefined;
+    replies[id] = undefined;
+  });
+
+  return { ...state, replies, inReplyTos };
+};
+
+const filterContexts = (
+  state: State,
+  relationship: ApiRelationshipJSON,
+  statuses: ImmutableList<Status>,
+) => {
+  const ownedStatusIds = statuses
+    .filter((status) => (status.get('account') as string) === relationship.id)
+    .map((status) => status.get('id') as string);
+
+  return deleteFromContexts(state, ownedStatusIds.toArray());
+};
+
+const updateContext = (state: State, status: ApiStatusJSON) => {
+  if (!status.in_reply_to_id) {
+    return state;
+  }
+
+  const replies = { ...state.replies };
+  const inReplyTos = { ...state.inReplyTos };
+  const siblings = (replies[status.in_reply_to_id] ??= []);
```

Comment:
Haven't we moved to a model where we can simply mutate the state in-place? If so, that may be preferable to re-defining things like this.